src/Misc/MiddleWare.cpp

```diff
@@ -266,6 +266,38 @@ static void idsFromMsg(const char* msg, int* part, int* kit, int* voice)
     }
 }
 
+/**
+ * Class responsible for sending chained requests of "/wavetable-rt-params"
+ */
+class WaveTableRequestHandler
+{
+    bool waitForAdPars[NUM_MIDI_PARTS][NUM_KIT_ITEMS][NUM_VOICES];
+
+public:
+    WaveTableRequestHandler()
+    {
+        for(std::size_t p = 0; p < NUM_MIDI_PARTS; ++p)
+        for(std::size_t k = 0; k < NUM_KIT_ITEMS; ++k)
+        for(std::size_t v = 0; v < NUM_VOICES; ++v)
+            waitForAdPars[p][k][v] = false;
+    }
+
+    void chainWtParamRequest(int part, int kit, int voice, rtosc::RtData& d)
+    {
+        if(!waitForAdPars[part][kit][voice])
+        {
+            std::string s = buildVoiceParMsg(part, kit, voice);
+            d.chain((s + "/wavetable-rt-params").c_str(), "");
+            waitForAdPars[part][kit][voice] = true;
```

Comment:
Perhaps I'm missing something, but I do worry that there's an ordering issue here. From my read of this commit, the first time there's a mutation needing the wavetable to update a message is fired off triggering the update. Other updates don't trigger another update while the existing one is pending. I guess the bit I don't see (though I'm guessing it's implemented elsewhere) is the stage where the first update is done and then the system sees "oh, this update is already stale" and then does the final update.

I'm guessing I'm just missing something though.

Reply:
Setting `waitForAdPars` does only stop MW from re-sending realtime parameter (currently only `Presonance`) requests.

Imagine 2 subsequent parameter changes are detected by MiddleWare, then the first change will set `waitForAdPars`, and request the realtime parameters. Now, if realtime parameters arrive at MW (port `request-wavetable:sTi:sFi`) before the second parameter, the wave is regenerated with only the first parameter changed, and `waitForAdPars` is set to `false`, leading the second parameter to re-request realtime parameters again (which will then trigger a "final update", as you say). If the second parameter comes in before realtime parameters, then the realtime parameter update at `request-wavetable:sTi:sFi` will trigger one wavetable regeneration with both parameters changed.

Does this sound correct to you?

Reply:
I think so. As long as the case has been considered then it's likely resolved.